---
title: "13_snRNAseq_diff_gene_expression_v1"
author: "Katie Prater"
date: "8/14/2020"
output: 
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Seurat)
library(Matrix)
library(ggplot2)
library(stringr)
library(dplyr)
library(filesstrings)
library(patchwork)
library(readr)
```

#Jayadev Lab snRNAseq Pipeline

##Detection of Differential genes by cluster

### R Markdown note:
This is an R Markdown document. When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document.

##First ensure you have the variables and directories set up as needed for the project. This should be identical to the code chunk you edited the first time. ***You DO need to edit this code chunk! ***
```{r Setup}
#Previous data file name: (give the name for the appopriate RDS/Rdata file)
start_data <- "QCd_norm_anchored_clustered_wscale_ws_MG_subset_0.2subset012356_MG_subset_20200910.rdata"
  #"QCd_data_subset_Dx_idents_ws_MG_subset_0.2subset012356_20200910.rdata"

#Resolution chosen from previous clustering to use to label clusters:
change_resolution <- "0.3"

#Change output path if necessary, otherwise leave NULL
change_output_path <- "../output/output_pu1_only/degs/MG_subset_0.2subset012356/"

#Analyze clusters by specific grouping (i.e. a metadata variable)? - Boolean
clust_group <- TRUE

#If clust_group is TRUE, variables in your metadata to group the cluster numbers by (must match variable names in metadata). Otherwise, NULL.
meta_vars <- c("Dx")

#__________________**DO NOT EDIT BELOW THIS LINE**___________________________
```

## Load in the data you specified:
```{r Load_Data}
#Document your code:
print("This is when this code was started:")
print(Sys.time())

#Load the rData saved before
print(paste('Loading data', start_data, sep = " "))
load(paste0("../data/r_files/", start_data))

#Change output path if a new one is provided.
if (!is.null(change_output_path)) {
  print("Changing output path")
  outpath <- change_output_path
}

#Ensure resolution to be used is set properly.
if (!is.null(change_resolution)) {
  chose_resolution <- change_resolution
}
```

```{r Set_Directories}
#Ensure paths are set for code:
if (file.exists(outpath))
    {cat("Output directory path works!")} else 
    {cat("Output directory does not exist - creating")
    dir.create(outpath)}

#Number of participants/samples to process:
numsubs <- length(samples)
print(paste("Processing", numsubs, "samples", sep = " "))

#Documentation:
print("This is where the project is:")
print(projdir)
print("This is where the sample files are:")
print(sample_dir)
print("This is where the data files will be saved:")
print(outpath)

```

## Calculate genes that differentiate groups from each other.
```{r}
##****** Not flexible!

# Find genes that are differentially expressed between groups.
print(paste0('Finding genes that differentiate groups defined by: ', meta_vars))
group.markers <- FindMarkers(ss_data_norm, assay = "RNA", group.by = meta_vars, ident.1 = "CTRL", test.use = "MAST", min.pct = 0.40, logfc.threshold = 0.25) 
head(group.markers, n=20)

#Write the genes to a csv file.
write.csv(group.markers, str_c(outpath,"/", proj, "_DEGs_group_min.4_", meta_vars, "_", chose_resolution, suffix, ".csv"), row.names = TRUE)

```

## Calculate genes that differentiate each cluster from all other clusters.
```{r Calc_Diff}
##First ensure that the idents are set to the resolution of clusters that you want.
#remake resolution name to ensure it's accurate:
res_name<- paste0(assay, "_snn_res.", chose_resolution)

#Set cluster identities in Seurat object to the numbers from the resolution you chose previously
Idents(object = ss_data_norm)<-res_name

# find markers for every cluster compared to all remaining cells to support cell identity verification. Report only the positive ones.
print('Finding genes that differentiate each cluster from other cells.')
data.markers <- FindAllMarkers(ss_data_norm, test.use = "MAST", only.pos = FALSE, min.pct = 0.25, logfc.threshold = 0.25) 

#Write the top cluster genes to a csv file.
write_csv(data.markers, str_c(outpath,"/", proj, "_top_genes_per_cluster_compared_all_res_", chose_resolution, suffix, ".csv"))

```
#Make a series of plots for the gene expression between clusters.
```{r Top3_Plot}
#Find the top 3 genes per cluster
top3_genes_cluster <- data.markers %>% group_by(cluster) %>% top_n(n = 3, wt = avg_logFC)
#make a list of them to plot
top3 <- unique(pull(top3_genes_cluster, gene))

#Make a dotplot of the top three genes for each cluster
plot1 <- DotPlot(ss_data_norm, 
                features = top3,
                assay = assay,
                split.by = var_subset,
                cols = c("blue", "violet")) + RotatedAxis() + theme(axis.text.x = element_text(size = 9, lineheight = 20))
ggsave(filename = paste0(outpath,"/", proj, "_top_genes_per_cluster_dotplot_", chose_resolution, suffix, ".pdf"), height = 7, width = 11, plot = plot1)

```

```{r Top10_Plots}
#Find the top 10 genes per cluster
top10_genes_cluster <- data.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
#make a list of them to plot
top10 <- unique(pull(top10_genes_cluster, gene))

#make resolution name:
      res_name<- paste0(assay, "_snn_res.", chose_resolution)

      #Set cluster identities in Seurat object
      Idents(object = ss_data_norm)<-res_name
      
      plot2 <- DoHeatmap(ss_data_norm, 
                        features = top10,
                        assay = assay,
                        raster = FALSE,
                        hjust = 1,
                        angle = 0)
      ggsave(filename = paste0(outpath,"/", cell_type, "_DEG_expression_cluster_heatmap_", chose_resolution, suffix, ".pdf"), height = 7, width = 11, plot = plot2)
      
for (clusters in 1:length(unique(Idents(ss_data_norm)))-1) {
  clust_genes <- data.markers %>% group_by(cluster) %>% filter(cluster == clusters) %>%  top_n(n = 10, wt = avg_logFC)
clust_gene_list <- pull(clust_genes, gene)
  
  #Plot the genes for this cluster on umaps
  plot3 <- FeaturePlot(ss_data_norm, 
            reduction = "umap", 
            features = clust_gene_list) +
        plot_annotation(title = paste0("DEGs for Cluster ", clusters))
     ggsave(filename = paste0(outpath,"/", cell_type, "_cluster_DEG_expression_umap_", chose_resolution, "_cluster_", clusters, suffix, ".png"), height = 7, width = 11, plot = plot3)

  #Make violin plots of the top genes by cluster.    
  for (plot in 1:10) {
       vln_plots <- VlnPlot(ss_data_norm, features = clust_gene_list[plot], split.by = var_subset,  pt.size = 0, combine = FALSE)
       plot4 <- wrap_plots(vln_plots)
       ggsave(filename = paste0(outpath,"/", cell_type, "_vln_plot_", chose_resolution, "_cluster_", clusters, "_", clust_gene_list[plot], suffix, ".png"), height = 7, width = 11, plot = plot4)  
  }
}
      

```

