---
title: "15_snRNAseq_interactive_plots_v1"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Seurat)
library(Matrix)
library(ggplot2)
library(stringr)
library(dplyr)
library(filesstrings)
library(cowplot)
library(plotly)
```

## Plots that allow hover or other functionality to explore the data in real-time. Do not run this by knitting.

```{r Plot_Diff_Genes}
#Set the theme of ggplot so that it makes the appropriate plots
theme_set(theme_cowplot())

#Subset the Seurat object and find average gene expression between the subset defined previously
microglia1 <- subset(ss_data_norm, idents = "Microglia 1")
Idents(microglia1) <- var_subset
avg.mic1.cells <- log1p(AverageExpression(microglia1, verbose = FALSE)$RNA)
avg.mic1.cells$gene <- rownames(avg.mic1.cells)

genes.to.label = unique(pull(slice_head(avg.mic1.cells, n=10), gene))
p1 <- ggplot(avg.mic1.cells, aes(CTRL, AD)) + geom_point() + ggtitle("Microglia 1 Cluster") + geom_abline()
HoverLocator(p1)


p1 <- LabelPoints(plot = p1, points = genes.to.label, repel = TRUE)
p2 <- ggplot(avg.cd14.mono, aes(CTRL, STIM)) + geom_point() + ggtitle("CD14 Monocytes")
p2 <- LabelPoints(plot = p2, points = genes.to.label, repel = TRUE)
plot_grid(p1, p2)

```

